pkg/network/certificates_knobs.go

```diff
@@ -0,0 +1,99 @@
+)
+
+// validateCertificateKnobs validates the following fields
+// - CARotateInterval
+// - CAOverlapInterval
+// - CertRotateInterval
+func validateCertificateKnobs(conf *opv1alpha1.NetworkAddonsConfigSpec) []error {
+	if conf.CertificateKnobs == nil {
+		return []error{}
+	}
+
+	certificateKnobs := *conf.CertificateKnobs
+
+	errs := []error{}
+
+	caRotateInterval := parseCertificateKnob("caRotateInterval", certificateKnobs.CARotateInterval, &errs)
+	caOverlapInterval := parseCertificateKnob("caOverlapInterval", certificateKnobs.CAOverlapInterval, &errs)
+	certRotateInterval := parseCertificateKnob("certRotateInterval", certificateKnobs.CertRotateInterval, &errs)
+
+	// If they cannot be parsed don't continue
+	if len(errs) > 0 {
+		return errs
+	}
+
+	validateGreaterThanZero("caRotateInterval", caRotateInterval, &errs)
+	validateGreaterThanZero("caOverlapInterval", caOverlapInterval, &errs)
+	validateGreaterThanZero("certRotateInterval", certRotateInterval, &errs)
+
+	// If we have a zero value don't continue
+	if len(errs) > 0 {
+		return errs
+	}
+
+	validateGreaterThan("caRotateInterval", caRotateInterval, "caOverlapInterval", caOverlapInterval, &errs)
+	validateGreaterThan("caRotateInterval", caRotateInterval, "certRotateInterval", certRotateInterval, &errs)
+
+	return errs
+}
+
+func fillDefaultsCertificateKnobs(conf, previous *opv1alpha1.NetworkAddonsConfigSpec) []error {
+	if conf.CertificateKnobs == nil {
+		return []error{}
+	}
+
+	if conf.CertificateKnobs.CARotateInterval == "" || conf.CertificateKnobs.CAOverlapInterval == "" || conf.CertificateKnobs.CertRotateInterval == "" {
+		if previous != nil && previous.CertificateKnobs != nil {
+			conf.CertificateKnobs = previous.CertificateKnobs
+		}
+	}
+	return []error{}
+}
+
+func changeSafeCertificateKnobs(prev, next *opv1alpha1.NetworkAddonsConfigSpec) []error {
+	if prev.CertificateKnobs != nil && next.CertificateKnobs != nil && !reflect.DeepEqual(prev.CertificateKnobs, next.CertificateKnobs) {
+		return []error{errors.Errorf("cannot modify CertificateKnobs configuration once it is deployed")}
+	}
+	return []error{}
+}
+
+func parseCertificateKnob(name, value string, errs *[]error) *time.Duration {
+	if value == "" {
+		return nil
+	}
+	d, err := time.ParseDuration(value)
+	if err != nil {
+		*errs = append(*errs, errors.Wrapf(err, "failed to parse %s", name))
+		return nil
+	}
+	return &d
+}
+
+func validateGreaterThanZero(name string, d *time.Duration, errs *[]error) {
+	if d == nil {
+		return
+	}
+	if *d == 0 {
+		*errs = append(*errs, errors.Errorf("failed to validate %s: duration has to be > 0", name))
+	}
+
+}
+
+func validateGreaterThan(lhsName string, lhsValue *time.Duration, rhsNamed string, rhsValue *time.Duration, errs *[]error) {
+	if lhsValue == nil || rhsValue == nil {
+		return
+	}
+
+	if *rhsValue > *lhsValue {
+		*errs = append(*errs, errors.Errorf("failed to validate: %s(%s) has to be <= %s(%s)", rhsNamed, *rhsValue, lhsName, *lhsValue))
```

Comment:
Return list of errors and join them on the caller level. We do it that way for all the other functions

Reply:
Done. But with a helper function.